ruby/rails/security/audit/xss/avoid-tainted-file-access.yaml

```diff
@@ -0,0 +1,61 @@
+rules:
+- id: avoid-tainted-file-access
+  metadata:
+    owasp: 'A6: Security Misconfiguration'
+    cwe: 'CWE-276: Incorrect Default Permissions'
+    references:
+    - https://github.com/presidentbeef/brakeman/blob/main/docs/warning_types/file_access/index.markdown
+    category: security
+    technology:
+    - rails
+  message: |
+    Using user input when accessing files is potentially dangerous. A malicious actor could use this to modify or access files they have no right to. [Not sure what to say for how to fix it]
```

Comment:
@minusworld @clintgibler Augment message

Reply:
Currently the way this is written assumes that all of the functions in `metavariable-pattern` for `$X` could happen for each of the modules: `Dir`, `File`, etc.

But in many cases, not all modules have all of those methods - for example [Net::FTP](https://docs.ruby-lang.org/en/2.5.0/Net/FTP.html) does not have `entries`, `glob`, and a bunch of other ones.

I think warning on any non string literal call to any of these function may be quite noisy in practice.

Might be a good use of taint mode to see if controller input (URL param, form param, cookie, ...) can reach one of these functions.

@minusworld Thoughts on splitting this rule out into multiple rules based on sink? Something like: shell exec (code injection), file I/O (arbitrary file read/write), and network (Net::FTP/Net::HTTP -> open redirect or SSRF or something)?

Reply:
1. Mm I was wondering this, because (if I read correctly) brakeman also took this approach of checking every combination, including ones that aren't possible? I do think it's easier and ok to do it this way, since it shouldn't cause false positives.
2. I can fix that! Now that I understand ruby/brakeman better, I see that I misinterpreted this